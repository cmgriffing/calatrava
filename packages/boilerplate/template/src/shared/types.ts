
export enum Tables {
  Users = "users",
ğŸ‘‰#hasTeamsğŸ‘ˆ
  Teams = "teams",
  Teammates = "teammates",
ğŸ‘‰/hasTeamsğŸ‘ˆ
}

export interface DatastoreRecord {
  createdAt: number;
  modifiedAt: number;
}

export interface User extends DatastoreRecord {
  userId: string;
  name: string;
  email: string;
  passwordHash: string;
  verificationToken: string;
  verified: boolean;
}

ğŸ‘‰#hasTeamsğŸ‘ˆ
export interface Team extends DatastoreRecord {
  teamId: string;
  ownerId: string;
  name: string;
}

export interface Teammate extends DatastoreRecord {
  teamId: string;
  userId: string;
  accepted: boolean;
  inviteToken: string;
}
ğŸ‘‰/hasTeamsğŸ‘ˆ

export interface TableTypes {
  [Tables.Users]: User;
ğŸ‘‰#hasTeamsğŸ‘ˆ
  [Tables.Teams]: Team;
  [Tables.Teammates]: Teammate;
ğŸ‘‰/hasTeamsğŸ‘ˆ
}
