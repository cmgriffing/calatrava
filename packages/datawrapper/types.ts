export enum DBKeys {
  Partition = "partitionKey",
  Sort = "sortKey",
  Tertiary = "tertiaryKey",
}

export interface Datastore {
  put: Function;
  get: Function;
  delete: Function;
  query: Function;
  scan: Function;
  update: Function;
  _doc: {
    batchGet: Function;
  };
}

export type BatchGetKeys = {
  [value in DBKeys]?: string;
};

export interface QueryKeys {
  [key: string]: string;
}

export interface GetAllOptions {
  filterExpression?: string;
  filterExpressionNames?: { [key: string]: string };
  filterExpressionValues?: { [key: string]: string };
  indexKey?: string;
}

export type DBRecord<T> =
  | T
  | { partitionKey: string; sortKey: string; tertiaryKey?: string };

export type CleanRecord<T> = Omit<
  T,
  "partitionKey" | "sortKey" | "tertiaryKey"
>;

export interface WrappedDatastore<T> {
  // GET_EVERYTHING: () => Promise<any>;
  create: (putObject: DBRecord<T>) => Promise<T>;
  getById: (
    idValue: QueryKeys,
    index?: DBKeys,
    secondaryId?: string
  ) => Promise<T>;
  getByIndex: (
    idValue: QueryKeys,
    index?: DBKeys,
    secondaryId?: string
  ) => Promise<T>;
  getRandom: (
    ignoreKey?: string,
    ignoreValue?: string,
    index?: DBKeys
  ) => Promise<T>;
  scanIdsByFilter: (options?: GetAllOptions) => Promise<T[]>;
  getAllById: (
    idValue: QueryKeys,
    options?: GetAllOptions,
    index?: DBKeys,
    secondaryId?: string
  ) => Promise<T[]>;
  getAllByManyIds: (
    idValues: string[],
    idKey: string,
    index?: DBKeys,
    secondaryId?: string
  ) => Promise<T[]>;
  update: (
    idValue: QueryKeys,
    patchObject: Partial<T>,
    secondaryId: QueryKeys,
    index?: DBKeys
  ) => Promise<T>;
  remove: (
    idValue: QueryKeys,
    secondaryId: QueryKeys,
    index?: DBKeys
  ) => Promise<T>;
}
