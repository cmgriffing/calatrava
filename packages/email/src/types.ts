import { AxiosInstance } from "axios";

export enum EmailProvider {
  Mailjet = "mailjet",
  Sendgrid = "sendgrid",
}

export interface EmailService {
  createAxios: (accessKey: string, extraKey?: string) => AxiosInstance;
  createSendEmail: (
    axios: AxiosInstance,
    emailTemplateIdMap: { [key: string]: string },
    fromEmail: string,
    debugMode: boolean
  ) => (toEmail: string, template: string, dynamicData: Object) => Promise<any>;

  createGetExistingTemplates: (
    axios: AxiosInstance
  ) => () => Promise<EmailTemplates>;

  processTemplate: (templateContent: string) => Promise<string>;

  createCreateNewTemplateIfNeeded: (
    axios: AxiosInstance
  ) => (
    existingTemplates: EmailTemplates,
    templateName: string
  ) => Promise<EmailTemplates>;

  createUploadCurrentContentIfNeeded: (
    axios: AxiosInstance
  ) => (
    existingTemplates: EmailTemplates,
    templateName: string,
    templateData: TemplateData,
    templateContent: string
  ) => Promise<void>;
}

export interface TemplateData {
  template: string;
  subject: string;
}

export interface EmailTemplate {
  id: string;
  version: string;
  content: string;
}

export interface EmailTemplates {
  [key: string]: EmailTemplate;
}
