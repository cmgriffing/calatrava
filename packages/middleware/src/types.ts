import { PresignedPost } from "@aws-sdk/s3-presigned-post";
import type { HttpRequest, HttpResponse } from "@architect/functions/http";
import { WrappedDatastore } from "@calatrava/datawrapper";

export interface TypedHttpResponse<T> extends HttpResponse {
  json?: T | undefined;
}

export interface BaseUser extends Object {
  userId: string;
}

export interface BaseTeam extends Object {
  teamId: string;
}

export interface BaseTeammate extends Object {
  teamId: string;
}

export interface HttpRequestWithTables extends HttpRequest {
  tables: {
    get: <T extends {}>(tableName?: string) => WrappedDatastore<T>;
  };
}

export interface HttpRequestWithUser<User extends BaseUser>
  extends HttpRequestWithTables {
  user: User;
}

export interface HttpRequestWithTeams<
  Team extends BaseTeam,
  User extends BaseUser
> extends HttpRequestWithUser<User> {
  ownedTeams: Team[];
  joinedTeams: Team[];
}

export interface HttpRequestWithSignableEntity<
  T,
  Team extends BaseTeam,
  User extends BaseUser
> extends HttpRequestWithTeams<Team, User> {
  entity: T;
}

export interface HttpRequestWithPresignedPost<
  T,
  Team extends BaseTeam,
  User extends BaseUser
> extends HttpRequestWithSignableEntity<T, Team, User> {
  presignedPost: PresignedPost;
}
