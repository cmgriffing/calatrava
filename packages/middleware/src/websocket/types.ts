import { HttpResponse } from "@architect/functions/http";
import { WrappedDatastore } from "@calatrava/datawrapper";
import { WebSocketIncomingMessage } from "./request-types";

export interface WebSocketRequest {
  requestContext: {
    connectedAt: number;
    connectionId: string;
    domainName: string;
    eventType: "MESSAGE";
    messageDirection: "IN" | "OUT";
    messageId: string;
    requestId: string;
    requestTimeEpoch: number;
    routeKey: string;
    stage: string;
  };
  isBase64Encoded: boolean;
  body: WebSocketIncomingMessage;
}

export interface WebSocketRequestWithTables extends WebSocketRequest {
  tables: {
    get: <T extends {}>(
      prop: string,
      tableName?: string
    ) => WrappedDatastore<T>;
  };
}

export interface WebSocketRequestWithUser<User>
  extends WebSocketRequestWithTables {
  user: User;
}

export type WebSocketHandler = (req: WebSocketRequest) => Promise<HttpResponse>;
