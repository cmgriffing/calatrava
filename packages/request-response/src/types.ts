import {
  HTTPClientErrorResponses,
  HTTPServerErrorResponses,
} from "./http-status";

export const commonHeaders = {
  "access-control-allow-origin": "*",
  "access-control-allow-methods": "GET, POST, PUT, PATCH, DELETE,OPTIONS",
  "access-control-allow-headers": "content-type,authorization",
  "cache-control": "no-cache, no-store, must-revalidate, max-age=0, s-maxage=0",
  "content-type": "application/json",
};

export interface RouteOptions {
  // used to determine which private paths get annotated with auth params
  open?: boolean;
  // used to determine which openapi-yaml file to include this route in
  public?: boolean;
  path: string;
  summary: string;
  description: string;
  tags: string[];
  headers: { [key: string]: string };
  method: "GET" | "POST" | "PUT" | "PATCH" | "DELETE" | "OPTIONS";
  definedErrors: (HTTPClientErrorResponses | HTTPServerErrorResponses)[];
  // Maybe we could combine the schema paths into a common interpolatable variable. (postUser)
  requestSchema?: string;
  responseSchema: string;
  errorSchema: string | "ErrorResponse";
}

// There is probably a better place for this
export function Route(_options: RouteOptions): Function {
  return () => {
    return;
  };
}
