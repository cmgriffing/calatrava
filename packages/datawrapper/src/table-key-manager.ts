import jsStringEscape from "js-string-escape";
import * as Case from "case";
import { DBKeys } from "./types";

export type TableKeysMap = {
  [key: string]: { [value in DBKeys]?: string[] };
};

export class TableKeyManager {
  private tableKeysMap: TableKeysMap;
  private tableNames: { [key: string]: string };

  constructor(
    tableKeysMap: TableKeysMap,
    tableNames: { [key: string]: string }
  ) {
    this.tableKeysMap = tableKeysMap;
    this.tableNames = tableNames;
  }

  private createKeyString(
    tableName: string,
    keyValueMap: { [key: string]: string | number | boolean }
  ): string {
    if (!Object.values(this.tableNames).includes(tableName)) {
      throw new Error(
        `Could not find table name, ${tableName}, in tableNames map.`
      );
    }

    let keyString = `#${Case.constant(tableName)}`;

    Object.entries(keyValueMap)
      .sort((entryA, entryB) => {
        if (entryA[0] > entryB[0]) {
          return 1;
        } else if (entryA[0] < entryB[0]) {
          return -1;
        } else {
          return 0;
        }
      })
      .forEach(([key, value]) => {
        keyString = `${keyString}#${Case.constant(
          jsStringEscape(key)
        )}#${jsStringEscape(value)}`;
      });

    return keyString;
  }

  getTable(table: string) {
    return {
      getTableKey: (
        dbKey: DBKeys,
        keyValueMap: { [key: string]: string | number | boolean }
      ) => {
        // validate keyValueMap
        const allowedKeyNames = this.tableKeysMap?.[table]?.[dbKey] || [];

        allowedKeyNames.forEach((keyName) => {
          if (!keyValueMap[keyName]) {
            throw new Error(
              `keyValueMap is missing required key: ${keyName}, table: ${table}, dbKey: ${dbKey}, keyValueMap: ${JSON.stringify(
                keyValueMap
              )}`
            );
          }
        });

        if (Object.keys(keyValueMap).length > allowedKeyNames.length) {
          console.warn(
            "Extra keys were passed in the keyValueMap. They will be ignored.",
            { allowedKeyNames, keyValueMap, table, dbKey }
          );
        }

        return this.createKeyString(table, keyValueMap);
      },
    };
  }
}

// end-user examples

// enum Tables {
//   foo = "foo",
// }

// const tables = arc.tables();

// const ourTableMeta = new TableKeyManager(
//   {
//     foo: {
//       [DBKeys.Partition]: ["userId"],
//     },
//   },
//   { Foo: "foo" }
// );

// const partitionKey = ourTableMeta
//   .getTable(Tables.foo)
//   .getTableKey(DBKeys.Partition, {
//     foo: "bar",
//   });
