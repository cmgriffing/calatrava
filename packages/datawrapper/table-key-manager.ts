import * as jsStringEscape from "js-string-escape";
import * as Case from "case";
import { DBKeys, QueryKeys } from "./types";

export type TableKeysMap = {
  [key: string]: { [value in DBKeys]?: string[] };
};

export class TableKeyManager {
  private tableKeysMap: TableKeysMap;
  private tableNames: { [key: string]: string };

  constructor(
    tableKeysMap: TableKeysMap,
    tableNames: { [key: string]: string }
  ) {
    this.tableKeysMap = tableKeysMap;
    this.tableNames = tableNames;
  }

  private createKeyString(
    tableName: string,
    keyValueMap: { [key: string]: string | number | boolean }
  ): string {
    if (!Object.values(this.tableNames).includes(tableName)) {
      throw new Error(
        `Could not find table name, ${tableName}, in tableNames map.`
      );
    }

    let keyString = `#${Case.constant(tableName)}`;

    Object.entries(keyValueMap).forEach(([key, value]) => {
      keyString = `${keyString}#${Case.constant(
        jsStringEscape(key)
      )}#${jsStringEscape(value)}`;
    });

    return keyString;
  }

  getTable(table: string) {
    return {
      getTableKey: (
        dbKey: DBKeys,
        keyValueMap: { [key: string]: string | number | boolean }
      ) => {
        // validate keyValueMap
        const allowedKeyNames = this.tableKeysMap[table][dbKey];

        allowedKeyNames.forEach((keyName) => {
          if (!keyValueMap[keyName]) {
            throw new Error(`keyValueMap is missing required key: ${keyName}`);
          }
        });

        if (Object.keys(keyValueMap).length > allowedKeyNames.length) {
          console.warn(
            "Extra keys were passed in the keyValueMap. They will be ignored.",
            { allowedKeyNames, keyValueMap }
          );
        }

        return this.createKeyString(table, keyValueMap);
      },
    };
  }
}

// end-user examples

// enum Tables {
//   foo = "foo",
// }

// const tables = arc.tables();

// const ourTableMeta = new TableKeyManager(
//   {
//     foo: {
//       [DBKeys.Partition]: ["userId"],
//     },
//   },
//   { Foo: "foo" }
// );

// const partitionKey = ourTableMeta
//   .getTable(Tables.foo)
//   .getTableKey(DBKeys.Partition, {
//     foo: "bar",
//   });
