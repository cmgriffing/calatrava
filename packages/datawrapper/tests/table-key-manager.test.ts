import { DBKeys } from "../types";
import { TableKeyManager } from "../table-key-manager";

enum Tables {
  foo = "foo",
}

const tableKeyManager = new TableKeyManager(
  {
    foo: {
      [DBKeys.Partition]: ["userId"],
    },
  },
  { Foo: "foo" }
);

test("Key manager generates a partition key", () => {
  const partitionKey = tableKeyManager
    .getTable(Tables.foo)
    .getTableKey(DBKeys.Partition, {
      userId: "abc123",
    });
  expect(partitionKey).toMatchSnapshot();
});
