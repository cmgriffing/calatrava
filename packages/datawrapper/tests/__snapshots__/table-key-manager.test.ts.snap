// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Key manager generates a partition key 1`] = `"#FOO#USER_ID#abc123"`;
